crates/tags/src/signed.rs

```diff
@@ -0,0 +1,242 @@
+                digest: format!("{}=:{}:", ALGORITHM, HASH).parse().unwrap(),
+            }),
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn tag_value_flattened_from_request() {
+        const SIGNATURE: &str = "DtEhU3ljbEg8L38VWAfUAqOyKAM6-Xx-F4GawxaepmXFCgfTjDxw5djxLa8ISlSApmWQxfKTUJqPP3-Kg6NU1Q";
+
+        let mut header = BTreeMap::new();
+        header.insert("kid".into(), "e9bc097a-ce51-4036-9562-d2ade882db0d".into());
+
+        let mut protected = BTreeMap::new();
+        protected.insert("alg".into(), "ES256".into());
+
+        test_from_request(
+            json!({
+                "payload": PAYLOAD,
+                "header": { "kid": "e9bc097a-ce51-4036-9562-d2ade882db0d" },
+                "protected": PROTECTED,
+                "signature": SIGNATURE,
+            }),
+            Kind::Signed(Jws::Flattened(Flattened {
+                payload: PAYLOAD.parse().unwrap(),
+                signature: Signature {
+                    header: Some(header),
+                    protected: Some(Json(protected)),
+                    signature: SIGNATURE.parse().unwrap(),
+                },
+            })),
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn tag_value_protected_from_request() {
+        const SIGNATURE: &str = "DtEhU3ljbEg8L38VWAfUAqOyKAM6-Xx-F4GawxaepmXFCgfTjDxw5djxLa8ISlSApmWQxfKTUJqPP3-Kg6NU1Q";
+        let mut header = BTreeMap::new();
+        header.insert("alg".into(), "ES256".into());
+
+        test_from_request(
+            json!({
+                "payload": PAYLOAD,
+                "protected": PROTECTED,
+                "signature": SIGNATURE,
+            }),
+            Kind::Signed(Jws::Flattened(Flattened {
+                payload: PAYLOAD.parse().unwrap(),
+                signature: Signature {
+                    protected: Some(Json(header)), // TODO: What MIME type is this supposed to be?
+                    header: None,
+                    signature: SIGNATURE.parse().unwrap(),
+                },
+            })),
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn tag_value_signatures_from_request() {
+        const SIGNATURE: &str = "DtEhU3ljbEg8L38VWAfUAqOyKAM6-Xx-F4GawxaepmXFCgfTjDxw5djxLa8ISlSApmWQxfKTUJqPP3-Kg6NU1Q";
+
+        let mut header = BTreeMap::new();
+        header.insert("alg".into(), "ES256".into());
+
+        test_from_request(
+            json!({
+                "payload": PAYLOAD,
+                "signatures": [
+                    {
+                        "protected": PROTECTED,
+                        "signature": SIGNATURE,
+                    }
+                ]
+            }),
+            Kind::Signed(Jws::General(General {
+                payload: PAYLOAD.parse().unwrap(),
+                signatures: vec![Signature {
+                    protected: Some(Json(header)), // TODO: What MIME type is this supposed to be?
+                    header: None,
```

Comment:
Which header ends up where? Why don't we have a test case, where both fields are populated?

Reply:
The function `tag_value_flattened_from_request` has header, protected, and signature defined.